### PR TITLE
 Combine the two managers

### PR DESCRIPTION
**Pull request by vidartf**
This PR:
- Changes the previous hotpot managers to only inherit the async manager (plain + mapping). This also reduces the different things that need testing.
- Makes the pooled manager also have the `max_kernels` trait (but we still keep the limited manager separate in case someone want just that, since it is quite simple).
- Merges in the behavior from the pooled manager by @maartenbreddels . 
- It tries to keep the number of config options small and unconflicting (i.e. it tries to make it hard to create a combination of config options that will be invalid).
- Puts in the standard Jupyter license
- Includes a paired down version of the nbclient code for having robust execution of update/init code.

TODO:
- [x] Make tests for the update/initialize code (including the client helper code).

### CODE REVIEW

File: hotpot_km/pooled.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    return await aw
+
+async def _await_then_kill(km, aw_id):
+    return await km.get_kernel(await aw_id).shutdown_kernel()
+
+class PooledKernelManager(LimitedKernelManager, AsyncMultiKernelManager):
+    max_kernels = Integer(0, config=True,
+        help="The maximum number of concurrent kernels",
+    )
+
```
Comment (maartenbreddels) [inline hotpot_km/pooled.py:36]:
```suggestion
```
We already get this from the LimitedKernelManager

File: hotpot_km/tests/test_initialize.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
+
+import asyncio
+from contextlib import asynccontextmanager
+import mock
```
Comment (maartenbreddels) [inline hotpot_km/tests/test_initialize.py:4]:
```suggestion
from unittest import mock
```
available from 3.3 https://docs.python.org/3.6/library/unittest.mock.html

**Comment by maartenbreddels** [comment]
Really confused about the following, which I get when using the Pooled Manager from voila:
```python
 File "/Users/maartenbreddels/miniconda3/envs/jdaviz-dev/lib/python3.7/site-packages/jinja2/asyncsupport.py", line 174, in auto_await
    return await value
  File "/Users/maartenbreddels/src/QuantStack/voila/voila/handler.py", line 169, in _jinja_kernel_start
    env=self.kernel_env,
  File "/Users/maartenbreddels/miniconda3/envs/jdaviz-dev/lib/python3.7/site-packages/nbclient/util.py", line 85, in ensure_async
    result = await obj
  File "/Users/maartenbreddels/miniconda3/envs/jdaviz-dev/lib/python3.7/site-packages/hotpot_km/pooled.py", line 141, in start_kernel
    return await fut
  File "/Users/maartenbreddels/miniconda3/envs/jdaviz-dev/lib/python3.7/site-packages/jupyter_client/multikernelmanager.py", line 468, in start_kernel
    await km.start_kernel(**kwargs)
  File "/Users/maartenbreddels/miniconda3/envs/jdaviz-dev/lib/python3.7/site-packages/jupyter_client/manager.py", line 575, in start_kernel
    self.kernel = await self._launch_kernel(kernel_cmd, **kw)
  File "/Users/maartenbreddels/miniconda3/envs/jdaviz-dev/lib/python3.7/site-packages/jupyter_client/manager.py", line 556, in _launch_kernel
    res = launch_kernel(kernel_cmd, **kw)
  File "/Users/maartenbreddels/miniconda3/envs/jdaviz-dev/lib/python3.7/site-packages/jupyter_client/launcher.py", line 134, in launch_kernel
    proc = Popen(cmd, **kwargs)
TypeError: __init__() got an unexpected keyword argument 'path'
```

Which is triggered because in Voila we call:
```
        kernel_id = await ensure_async(self.kernel_manager.start_kernel(
           kernel_name=nb.metadata.kernelspec.name,
           path=self.cwd,
           env=self.kernel_env,
        ))
```

Which is normally accepted here: https://github.com/jupyter-server/jupyter_server/blob/36218db5f08b08e41f1cff36f40a29adb146f81e/jupyter_server/services/kernels/kernelmanager.py#L165

But not accepted by https://github.com/jupyter/jupyter_client/blob/987f3e1de21e34d7f6f8b3225a6b044e1bdb0152/jupyter_client/multikernelmanager.py#L463 and simply passed down via kwargs, leading to the issue.

We may want to copy this behaviour (https://github.com/jupyter-server/jupyter_server/blob/36218db5f08b08e41f1cff36f40a29adb146f81e/jupyter_server/services/kernels/kernelmanager.py#L182) to be compatible, what do you think? https://github.com/jupyter-server/jupyter_server/commit/cb25736a24efd06be1ebd4f75d64b35f2cc25b5a seems to be the commit that added the `path` argument.

**Comment by maartenbreddels** [comment]
Since https://github.com/jupyter-server/jupyter_server/blob/36218db5f08b08e41f1cff36f40a29adb146f81e/jupyter_server/utils.py#L105 means path is interpreted as API path, not a filesystem path, I think voila is wrong here, and voila should just use cwd.

File: hotpot_km/tests/test_initialize.py (hunk 2 above)
Comment (vidartf) [reply hotpot_km/tests/test_initialize.py:4]: Ah, I got them reversed.

**Comment by maartenbreddels** [comment]
Are you ok with inheriting from `jupyter_server.services.kernels.kernelmanager.AsyncMappingKernelManager` so we can use the culling features, and we can support the `path` argument? It does add it as a dependency, but that should be ok.

**Comment by vidartf** [comment]
Note that the `PooledMappingKernelManager` is available directly, but as an optional dependency. I'm assuming you mean to make it a direct dependency, and be the main export of the module.

**Comment by vidartf** [comment]
PS: I think the mapping manager's culler will cull pooled kernels the way the code is currently written ?

**Comment by maartenbreddels** [comment]
> Note that the `PooledMappingKernelManager` is available directly, but as an optional dependency.

Ah, I missed that! That might solve my problem.

**Comment by vidartf** [comment]
Merging to iterate

**Pull request merged**